File: frontend/src/utils/validation.ts
import type { ChatGenerateResponse } from "../types/chat";

export type UiValidationItem = {
  key: string;
  label: string;
  description: string;
  ok: boolean;
};

const VALIDATION_META: Record<string, { label: string; description: string }> = {
  non_empty: {
    label: "Ответ не пустой",
    description: "Модель действительно вернула содержательный результат, а не пустую строку.",
  },
  policy: {
    label: "Соответствует правилам безопасности",
    description: "В коде не обнаружены запрещённые конструкции и опасные API.",
  },
  syntax: {
    label: "Синтаксис Lua корректный",
    description: "Код проходит синтаксическую проверку и может быть разобран интерпретатором Lua.",
  },
  domain: {
    label: "Соответствует условиям задачи",
    description: "Код выглядит уместным для постановки задачи и не противоречит доменным ограничениям.",
  },
  runtime: {
    label: "Исполняется без ошибок",
    description: "Код проходит runtime-проверку и не падает при запуске в тестовой обёртке.",
  },
  scenario: {
    label: "Проходит сценарную проверку",
    description: "Решение похоже на ожидаемый шаблон для данного класса задач.",
  },
};

export function mapValidation(meta: ChatGenerateResponse | null): {
  isValid: boolean;
  items: UiValidationItem[];
  errors: string[];
  warnings: string[];
} | null {
  if (!meta?.validation) return null;

  const { validation } = meta;

  const items: UiValidationItem[] = Object.entries(
    validation.stage_results || {},
  ).map(([key, value]) => ({
    key,
    label: VALIDATION_META[key]?.label ?? key,
    description:
      VALIDATION_META[key]?.description ??
      "Системная проверка этого этапа выполнена без дополнительного описания.",
    ok: Boolean(value),
  }));

  return {
    isValid: validation.is_valid,
    items,
    errors: validation.errors || [],
    warnings: validation.warnings || [],
  };
}

File: localscript_backend/app/schemas/validation.py
from pydantic import BaseModel, Field


class ValidateCodeRequest(BaseModel):
    code: str = Field(
        ...,
        min_length=1,
        description="Lua-код для проверки",
    )
    task: str = Field(
        default="",
        description="Необязательный исходный текст задачи для доменной валидации",
    )
    validate_runtime: bool = Field(
        default=True,
        description="Нужно ли выполнять runtime-валидацию через lua",
    )


class ValidateCodeResponse(BaseModel):
    is_valid: bool = Field(..., description="Прошла ли валидация")
    stage_results: dict[str, bool] = Field(
        default_factory=dict,
        description="Результаты отдельных этапов валидации",
    )
    errors: list[str] = Field(
        default_factory=list,
        description="Ошибки валидации",
    )
    warnings: list[str] = Field(
        default_factory=list,
        description="Предупреждения валидации",
    )